File: tests/integration/index-types-test/index-types-test.cpp
#include "index-types-test.h"

using namespace js;

template<typename T, typename K>
js::any getProperty(T obj, K key) {
    return obj->key;
}

const Person person = []() {
      js::object obj_temp_0;
      obj_temp_0.set("name", "John"_S);
      obj_temp_0.set("age", js::number(30));
      obj_temp_0.set("email", "john@example.com"_S);
      return js::any(obj_temp_0);
    }();

const js::any name = getProperty(person, "name"_S);

const js::any age = getProperty(person, "age"_S);

const StringDictionary dict = []() {
      js::object obj_temp_1;
      obj_temp_1.set("hello", "world"_S);
      obj_temp_1.set("foo", "bar"_S);
      obj_temp_1.set("baz", "qux"_S);
      return js::any(obj_temp_1);
    }();

const NumberDictionary arr = []() {
      js::object obj_temp_2;
      obj_temp_2.set("0", "first"_S);
      obj_temp_2.set("1", "second"_S);
      obj_temp_2.set("2", "third"_S);
      return js::any(obj_temp_2);
    }();

const MixedDictionary mixed = []() {
      js::object obj_temp_3;
      obj_temp_3.set("length", js::number(10));
      obj_temp_3.set("name", "MyDictionary"_S);
      obj_temp_3.set("key1", "value1"_S);
      obj_temp_3.set("key2", js::number(42));
      return js::any(obj_temp_3);
    }();

template<typename T, typename K>
Pick pickProperties(T obj, js::array<K> keys) {
    const js::any result = []() {
          js::object obj_temp_4;
          return js::any(obj_temp_4);
        }();
    for (const auto& key : keys) {
            {
                        result->key = obj->key;
            }
    }
    return result;
}

const js::any nameAndAge = pickProperties(person, js::array<js::string>{"name"_S, "age"_S});

template<typename T, typename K>
js::array<js::any> pluck(js::array<T> objects, K key) {
    return objects.map([](auto obj) -> auto { return obj->key; });
}

js::array<Person> people = js::array<js::any>{[]() {
      js::object obj_temp_5;
      obj_temp_5.set("name", "Alice"_S);
      obj_temp_5.set("age", js::number(25));
      obj_temp_5.set("email", "alice@example.com"_S);
      return js::any(obj_temp_5);
    }(), []() {
      js::object obj_temp_6;
      obj_temp_6.set("name", "Bob"_S);
      obj_temp_6.set("age", js::number(30));
      obj_temp_6.set("email", "bob@example.com"_S);
      return js::any(obj_temp_6);
    }(), []() {
      js::object obj_temp_7;
      obj_temp_7.set("name", "Charlie"_S);
      obj_temp_7.set("age", js::number(35));
      obj_temp_7.set("email", "charlie@example.com"_S);
      return js::any(obj_temp_7);
    }()};

const js::any names = pluck(people, "name"_S);

const js::any ages = pluck(people, "age"_S);

// Entry point
void Main() {
    dict["newKey"_S] = "newValue"_S;
    js::console.log(dict["hello"_S]);
    js::console.log(arr[js::number(0)]);
    js::console.log(arr[js::number(1)]);
    js::console.log(mixed.length());
    js::console.log(mixed["key1"_S]);
    js::console.log(nameAndAge->name);
    js::console.log(nameAndAge->age);
    js::console.log("Names:"_S, names);
    js::console.log("Ages:"_S, ages);
    js::console.log("Index types test completed"_S);
}

int main(int /*argc*/, char** /*argv*/) {
    Main();
    return 0;
}

File: tests/integration/index-types-test.ts
// Index Types and Indexed Access Test
// Tests TypeScript's index types and indexed access features

// Basic indexed access types
interface Person {
  name: string;
  age: number;
  email: string;
}

// Indexed access - T[K]
type PersonName = Person["name"]; // string
type PersonAge = Person["age"]; // number
type PersonProperty = Person["name" | "age"]; // string | number

// Generic indexed access
function getProperty<T, K extends keyof T>(obj: T, key: K): T[K] {
  return obj[key];
}

// Index signatures
interface StringDictionary {
  [key: string]: string;
}

interface NumberDictionary {
  [index: number]: string;
}

// Mixed index signatures
interface MixedDictionary {
  [key: string]: string | number;
  length: number; // Special property
  name: string; // Special property
}

// Using index types
const person: Person = {
  name: "John",
  age: 30,
  email: "john@example.com",
};

const _name = getProperty(person, "name"); // string
const _age = getProperty(person, "age"); // number

// String index signature usage
const dict: StringDictionary = {
  hello: "world",
  foo: "bar",
  baz: "qux",
};

dict["newKey"] = "newValue";
console.log(dict["hello"]);

// Number index signature usage
const arr: NumberDictionary = {
  0: "first",
  1: "second",
  2: "third",
};

console.log(arr[0]);
console.log(arr[1]);

// Mixed dictionary usage
const mixed: MixedDictionary = {
  length: 10,
  name: "MyDictionary",
  key1: "value1",
  key2: 42,
};

console.log(mixed.length);
console.log(mixed["key1"]);

// Index types with mapped types
type Nullable<T> = {
  [K in keyof T]: T[K] | null;
};

type NullablePerson = Nullable<Person>;
// Results in:
// {
//   name: string | null;
//   age: number | null;
//   email: string | null;
// }

// Pick utility using index types
type Pick<T, K extends keyof T> = {
  [P in K]: T[P];
};

type PersonNameAndAge = Pick<Person, "name" | "age">;
// Results in:
// {
//   name: string;
//   age: number;
// }

// Function to demonstrate Pick-like behavior
function pickProperties<T, K extends keyof T>(obj: T, keys: K[]): Pick<T, K> {
  const result = {} as Pick<T, K>;
  for (const key of keys) {
    result[key] = obj[key];
  }
  return result;
}

const nameAndAge = pickProperties(person, ["name", "age"]);
console.log(nameAndAge.name);
console.log(nameAndAge.age);

// Index access with arrays
type ArrayElement<T> = T extends (infer U)[] ? U : never;
type StringArrayElement = ArrayElement<string[]>; // string

// Tuple indexed access
type Tuple = [string, number, boolean];
type FirstElement = Tuple[0]; // string
type SecondElement = Tuple[1]; // number
type TupleLength = Tuple["length"]; // 3

// Using index types in practice
function pluck<T, K extends keyof T>(objects: T[], key: K): T[K][] {
  return objects.map((obj) => obj[key]);
}

const people: Person[] = [
  { name: "Alice", age: 25, email: "alice@example.com" },
  { name: "Bob", age: 30, email: "bob@example.com" },
  { name: "Charlie", age: 35, email: "charlie@example.com" },
];

const names = pluck(people, "name"); // string[]
const ages = pluck(people, "age"); // number[]

console.log("Names:", names);
console.log("Ages:", ages);

console.log("Index types test completed");
